### PR TITLE
UIOR-218 UI validation is unsynced with back-end for location quantities

### PR DESCRIPTION
**Pull request by AndreiKandratsenka**
## Purpose
it needs to sync UI and back-end validation for location quantities
https://issues.folio.org/browse/UIOR-218

## Approach
created ReapeatableFieldCatchError for providing FieldArray errors,
created validator is location should be specified,
updated validators for location quantities fields

## Pre-Merge Checklist
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/components/POLine/Location/LocationForm.js [hunk 1]
```diff
@@ -27,33 +33,53 @@ const validateNotNegative = (value) => (value < 0
...
   return value ? Number(value) : 0;
 };
 
+export const isLocationsRequired = (values, valuesAll) => {
+  const orderFormat = get(valuesAll, 'orderFormat');
+  const physicalInventory = get(valuesAll, 'physical.createInventory', '');
+  const eresourceInventory = get(valuesAll, 'eresource.createInventory', '');
+
+  if (
+    (isLocationPhysicalQuantityRequired(orderFormat, physicalInventory)
```
Comment (aliaksei-chumakou) [inline src/components/POLine/Location/LocationForm.js]: something wrong with line breaks, IMHO 
```
if (
  (
    isLocationPhysicalQuantityRequired(orderFormat, physicalInventory) 
    || isLocationEresourceQuantityRequired(orderFormat, eresourceInventory)
  ) && !values.length
) {
```

Comment (aliaksei-chumakou) [reply src/components/POLine/Location/LocationForm.js]: but I suggest to extract booleans to readable vars, so condition could be read easily

File: src/components/Utils/RepeatableFieldCatchError/RepeatableFieldCatchError.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import React from 'react';
+import { get } from 'lodash';
+import { FormattedMessage } from 'react-intl';
+import { RepeatableField } from '@folio/stripes-components';
+import css from './RepeatableFieldCatchError.css';
+
+export const RepeatableFieldCatchError = (props) => { // eslint-disable-line import/prefer-default-export
```
Comment (aliaksei-chumakou) [inline src/components/Utils/RepeatableFieldCatchError/RepeatableFieldCatchError.js]: `RepeatableFieldCatchError.js` - is not the best name. You actually don't catch an error, you've added a block with error message. 
Also please use another dir for common code - it's src/common

File: src/components/POLine/const.js [hunk 3]
```diff
@@ -45,3 +45,22 @@ export const INVENTORY_RECORDS_TYPE = {
...
 };
+
+export const REQUIRED_LOCATION_QUANTITY_MAP_TO_INVENTORY = {
+  [INVENTORY_RECORDS_TYPE.all]: true,
+  [INVENTORY_RECORDS_TYPE.instance]: false,
+  [INVENTORY_RECORDS_TYPE.instanceAndHolding]: true,
+  [INVENTORY_RECORDS_TYPE.none]: false,
+};
+
+export const isLocationPhysicalQuantityRequired = (orderFormat, inventory) => (
```
Comment (aliaksei-chumakou) [inline src/components/POLine/const.js]: I'm not sure that file `const.js` should contain such service functions. IMHO, it's better to move their code to something like `validation.js` along with LocationForm files

**Comment by aliaksei-chumakou** [comment]
@AndreiKandratsenka A huge job, well done so far! However, some code clean-up could be performed, let's improve this, thanks

File: src/components/POLine/Location/validate.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+const LOCATION_MUST_BE_SPECIFIED = 'required';
+
+const getTotalLocationsQuantities = (locations, propName) => {
+  const reducer = (accumulator, d) => accumulator + (d[propName] || 0);
+
+  return locations.reduce(reducer, 0);
+};
+
+export const validateNotNegative = (value) => (value < 0
```
Comment (aliaksei-chumakou) [inline src/components/POLine/Location/validate.js]: this code style is confusing. We do have an eslint rule and you make code harder to read just to bypass it. Please remove braces and format the code as eslint suggests

Comment (aliaksei-chumakou) [inline src/components/POLine/Location/validate.js]: basically it ruins one of the main idea of arrow function - to make code style short and clean. If you can't make it on one line - just break it on several using regular `return` and function body.

File: src/common/RepeatableFieldWithErrorMessage/RepeatableFieldWithErrorMessage.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+import { RepeatableField } from '@folio/stripes-components';
+import css from './RepeatableFieldWithErrorMessage.css';
+
+export const RepeatableFieldWithErrorMessage = (props) => { // eslint-disable-line import/prefer-default-export
+  const error = get(props, 'meta.error');
+
+  return (
+    <React.Fragment>
+      {error &&
+      <div className={css.feedbackError}>
```
Comment (NikitaSedyx) [inline src/common/RepeatableFieldWithErrorMessage/RepeatableFieldWithErrorMessage.js]: code style please

File: src/common/RepeatableFieldWithErrorMessage/RepeatableFieldWithErrorMessage.js [hunk 6]
```diff
@@ -0,0 +1,19 @@
+import React from 'react';
+import { get } from 'lodash';
+import { FormattedMessage } from 'react-intl';
+import { RepeatableField } from '@folio/stripes-components';
+import css from './RepeatableFieldWithErrorMessage.css';
+
+export const RepeatableFieldWithErrorMessage = (props) => { // eslint-disable-line import/prefer-default-export
```
Comment (NikitaSedyx) [inline src/common/RepeatableFieldWithErrorMessage/RepeatableFieldWithErrorMessage.js]: will be nice to propose it to stripes team one day

File: src/components/POLine/Location/validate.js [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+export const REQUIRED_LOCATION_QUANTITY_MAP_TO_INVENTORY = {
+  [INVENTORY_RECORDS_TYPE.all]: true,
+  [INVENTORY_RECORDS_TYPE.instance]: false,
+  [INVENTORY_RECORDS_TYPE.instanceAndHolding]: true,
+  [INVENTORY_RECORDS_TYPE.none]: false,
+};
+
+export const isLocationPhysicalQuantityRequired = (orderFormat, inventory) => {
+  return [...PHRESOURCES, OTHER].includes(orderFormat)
+    ? REQUIRED_LOCATION_QUANTITY_MAP_TO_INVENTORY[inventory]
```
Comment (NikitaSedyx) [inline src/components/POLine/Location/validate.js]: whe should follow one pattern: operator on next line or on previous, I see we use both now

File: src/components/POLine/Location/validate.js [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+
+  if ((isPhysicalQuantityRequired || isElectronicQuantityRequired) && !valuesAll.locations.length) {
+    return LOCATION_MUST_BE_SPECIFIED;
+  }
+
+  return undefined;
+};
+
+export const validateLocation = (value, valuesAll) => {
+  const { quantityPhysical, quantityElectronic } = valuesAll.locations.find(({ locationId }) => locationId === value);
```
Comment (NikitaSedyx) [inline src/components/POLine/Location/validate.js]: break the line please

File: src/common/RepeatableFieldWithErrorMessage/RepeatableFieldWithErrorMessage.js (hunk 5 above)
Comment (AndreiKandratsenka) [reply src/common/RepeatableFieldWithErrorMessage/RepeatableFieldWithErrorMessage.js]: fixed

File: src/components/POLine/Location/validate.js (hunk 8 above)
Comment (AndreiKandratsenka) [reply src/components/POLine/Location/validate.js]: fixed

File: src/components/POLine/Location/validate.js (hunk 7 above)
Comment (aliaksei-chumakou) [reply src/components/POLine/Location/validate.js]: I think this one is correct

File: src/components/POLine/Location/LocationForm.js [hunk 9]
```diff
@@ -5,55 +5,30 @@ import {
...
-
-  return allLocationsQuantity === overallLineQuantity
-    ? undefined
-    : <FormattedMessage id="ui-orders.location.quantityElectronic.notMatch" />;
-};
-
-const parseQuantity = (value) => {
-  return value ? Number(value) : 0;
-};
+import { RepeatableFieldWithErrorMessage } from '../../../common/RepeatableFieldWithErrorMessage/RepeatableFieldWithErrorMessage';
```
Comment (AndreiShumski2) [inline src/components/POLine/Location/LocationForm.js]: It should be above import getLocationsForSelect

**Comment by craigmcnally** [comment]
Seems like there are a lot of code format back and forth... Do "we" use prettier?  js-beautify?  anything?  It would probably save a lot of time to just all adopt a standard.

File: src/common/RepeatableFieldWithErrorMessage/RepeatableFieldWithErrorMessage.js (hunk 6 above)
Comment (NikitaSedyx) [reply src/common/RepeatableFieldWithErrorMessage/RepeatableFieldWithErrorMessage.js]: actually I don't see reason you didn't use default export

**Pull request merged**